File: sim_verify/shared/math.js
/**
 * 星空策略共享数学库 (JavaScript)
 * 被 dashboard.html 和 trend-analyzer.html 共享
 */

function polyval(coeffs, x) {
  let y = 0;
  for (let i = 0; i < coeffs.length; i++) y = y * x + coeffs[i];
  return y;
}

function polyfit(x, y, degree) {
  const n = x.length, m = degree + 1;
  const A = Array.from({ length: n }, (_, i) =>
    Array.from({ length: m }, (_, j) => Math.pow(x[i], j))
  );
  const AT = Array.from({ length: m }, (_, j) =>
    Array.from({ length: n }, (_, i) => A[i][j])
  );
  const ATA = Array.from({ length: m }, (_, j) =>
    Array.from({ length: m }, (_, k) =>
      AT[j].reduce((s, v, i) => s + v * A[i][k], 0)
    )
  );
  const ATy = Array.from({ length: m }, (_, j) =>
    AT[j].reduce((s, v, i) => s + v * y[i], 0)
  );
  const aug = ATA.map((row, j) => [...row, ATy[j]]);
  for (let col = 0; col < m; col++) {
    let pivot = col;
    for (let r = col + 1; r < m; r++)
      if (Math.abs(aug[r][col]) > Math.abs(aug[pivot][col])) pivot = r;
    [aug[col], aug[pivot]] = [aug[pivot], aug[col]];
    const piv = aug[col][col];
    if (Math.abs(piv) < 1e-14) continue;
    for (let c = col; c <= m; c++) aug[col][c] /= piv;
    for (let r = 0; r < m; r++) {
      if (r === col) continue;
      const f = aug[r][col];
      for (let c = col; c <= m; c++) aug[r][c] -= f * aug[col][c];
    }
  }
  return aug.map(r => r[m]).reverse();
}

function randn() {
  let u = 0;
  while (u === 0) u = Math.random();
  return Math.sqrt(-2 * Math.log(u)) * Math.cos(2 * Math.PI * Math.random());
}

function sharpe(returns) {
  if (returns.length < 2) return -10;
  const mean = returns.reduce((a, b) => a + b, 0) / returns.length;
  const variance = returns.reduce((s, r) => s + (r - mean) ** 2, 0) / returns.length;
  if (variance < 1e-15) return -10;
  return (mean / Math.sqrt(variance)) * Math.sqrt(252);
}

function maxDrawdown(returns) {
  let cum = 1, peak = 1, maxDD = 0;
  for (const r of returns) {
    cum *= (1 + r);
    if (cum > peak) peak = cum;
    const dd = (cum - peak) / peak;
    if (dd < maxDD) maxDD = dd;
  }
  return maxDD;
}

function sigmoid(residual, k = 2.0, b = -0.74) {
  const arg = Math.max(Math.min(-k * Math.abs(residual) + b, 50), -50);
  return 1 / (1 + Math.exp(-arg));
}

function dynamicThreshold(holdingTime, sigma0 = 0.5, tau = 5.0) {
  const lam = Math.log(2) / tau;
  return sigma0 * Math.exp(-lam * holdingTime);
}

function Cnk(n, k) {
  if (k > n || k < 0) return 0;
  if (k === 0 || k === n) return 1;
  let r = 1;
  for (let i = 1; i <= k; i++) r = r * (n - i + 1) / i;
  return Math.round(r);
}

function cubicDiscriminant(coeffs) {
  const [a, b, c, d] = coeffs;
  if (Math.abs(a) < 1e-15) return { p: 0, q: 0, delta: 0, type: 'degenerate' };
  const p = (3 * a * c - b * b) / (3 * a * a);
  const q = (2 * b * b * b - 9 * a * b * c + 27 * a * a * d) / (27 * a * a * a);
  const delta = (q / 2) ** 2 + (p / 3) ** 3;
  let type;
  if (delta > 1e-10) type = '1_real_root';
  else if (Math.abs(delta) < 1e-10) type = 'multiple_roots';
  else type = '3_real_roots';
  return { p, q, delta, type };
}

// Zone determination
function getZone(holdingTime, tau) {
  if (holdingTime < tau) return 1;
  if (holdingTime < 2 * tau) return 2;
  return 3;
}

// EMA smoothing
function emaSmooth(values, tau) {
  const alpha = 1 - Math.pow(2, -1 / tau);
  const smoothed = [values[0]];
  for (let i = 1; i < values.length; i++)
    smoothed.push(alpha * values[i] + (1 - alpha) * smoothed[i - 1]);
  return smoothed;
}

File: sim_verify/shared/math.py
"""
星空策略共享数学库 (Python)
所有 verify_*.py 工具的统一数学基础
"""
import numpy as np

def polyval(coeffs, x):
    """多项式求值"""
    return np.polyval(coeffs, x)

def polyfit(x, y, degree):
    """最小二乘多项式拟合"""
    return np.polyfit(x, y, degree)

def sharpe(returns, annualize=True):
    """夏普比 (年化)"""
    if len(returns) < 2:
        return -10.0
    mean = np.mean(returns)
    std = np.std(returns)
    if std < 1e-15:
        return -10.0
    s = mean / std
    return s * np.sqrt(252) if annualize else s

def sortino(returns, annualize=True):
    """Sortino比 (仅惩罚下行波动)"""
    if len(returns) < 2:
        return -10.0
    mean = np.mean(returns)
    downside = np.std([r for r in returns if r < 0]) if any(r < 0 for r in returns) else 1e-10
    s = mean / max(downside, 1e-10)
    return s * np.sqrt(252) if annualize else s

def max_drawdown(returns):
    """最大回撤"""
    if len(returns) == 0:
        return -10.0
    cumulative = np.cumprod(1 + np.array(returns))
    running_max = np.maximum.accumulate(cumulative)
    drawdowns = (cumulative - running_max) / running_max
    return float(np.min(drawdowns))

def sigmoid(residual, k=2.0, b=-0.74):
    """SIGMOID自信度函数: α(Δ) = σ(-k|Δ|+b)"""
    arg = -k * abs(residual) + b
    arg = max(min(arg, 50), -50)
    return 1.0 / (1.0 + np.exp(-arg))

def dynamic_threshold(holding_time, sigma0=0.5, tau=5.0):
    """动态阈值: σ(t) = σ₀·e^(-λt)"""
    lam = np.log(2) / tau
    return sigma0 * np.exp(-lam * holding_time)

def randn():
    """标准正态随机数"""
    return np.random.normal()

def generate_garch_noise(n, omega=0.01, alpha=0.1, beta=0.85):
    """生成 GARCH(1,1) 噪声序列"""
    sigma2 = np.zeros(n)
    eps = np.zeros(n)
    sigma2[0] = omega / (1 - alpha - beta)
    for t in range(1, n):
        sigma2[t] = omega + alpha * eps[t-1]**2 + beta * sigma2[t-1]
        eps[t] = np.sqrt(sigma2[t]) * randn()
    return eps

def generate_student_t_noise(n, nu=3):
    """生成 Student's t 分布噪声"""
    z = np.random.normal(size=n)
    chi2 = np.random.chisquare(nu, size=n)
    return z * np.sqrt(nu / chi2)

def cubic_discriminant(coeffs):
    """计算三次多项式的卡尔达诺判别式
    coeffs = [a, b, c, d] 对应 P(t) = a·t³ + b·t² + c·t + d
    返回 (p, q, delta, root_type)
    """
    a, b, c, d = coeffs[0], coeffs[1], coeffs[2], coeffs[3]
    if abs(a) < 1e-15:
        return 0, 0, 0, "degenerate"
    p = (3*a*c - b**2) / (3*a**2)
    q = (2*b**3 - 9*a*b*c + 27*a**2*d) / (27*a**3)
    delta = (q/2)**2 + (p/3)**3
    if delta > 0:
        return p, q, delta, "1_real_root"
    elif abs(delta) < 1e-10:
        return p, q, delta, "multiple_roots"
    else:
        return p, q, delta, "3_real_roots"
